Automattic/newspack-plugin
# feat(newspack-ui): integrate with sign in/sign up

### All Submissions:






### Changes proposed in this Pull Request:

This PR implements the Newspack UI  to the Sign in/Sign up modal and introduces changes to allow the modal integration with other components from this project.

The modal logic has been decoupled from the authentication form logic, allowing the modal management to be more easily extended to support callbacks and other customizations.

The new modal logic exposes an `openAuthModal( config )` to the reader activation library, which can be called by other components that can run a callback (another modal).

The configuration allows for the following customizations:

 - callback(message, data): function called after successful login
 - initialState: whether to open auth modal as `register` or `signin`
 - skipSuccess: whether to skip the final "success" state after login, close the modal and execute the callback immediately
 - labels: object containing custom labels for the auth form (modal title and button labels)
 - content: custom HTML to appear in the auth modal before the form

Asana tasks:
- 1205234045751551-as-1205668937699522/f
- 1205234045751551-as-1206405274735562/f
- 1205234045751551-as-1205668937699526/f

#### Sign In

<img width="404" alt="image" src="https://github.com/Automattic/newspack-plugin/assets/820752/41814832-3ccd-4e79-b939-c37513c6ada6">

#### Password

<img width="406" alt="image" src="https://github.com/Automattic/newspack-plugin/assets/820752/74ab12f9-e5a8-4fa1-b38e-6a075702d22e">

#### OTP

<img width="401" alt="image" src="https://github.com/Automattic/newspack-plugin/assets/820752/1e7a2062-eed3-48a9-81eb-e784a97c3dc5">

#### Create an account

<img width="399" alt="image" src="https://github.com/Automattic/newspack-plugin/assets/820752/ca251b15-7a17-4645-a2d3-ef4789d68a80">

#### Successful sign in

<img width="400" alt="image" src="https://github.com/Automattic/newspack-plugin/assets/820752/3af99a01-53a3-4c91-8fab-eb7fa5b87b5d">

#### Successful registration

<img width="399" alt="image" src="https://github.com/Automattic/newspack-plugin/assets/820752/308b51cc-e603-487a-9590-e00863329d2b">

#### Example of the `openAuthModal( config )` for the modal checkout

The example below is drafted in https://github.com/Automattic/newspack-blocks/pull/1647

<img width="390" alt="image" src="https://github.com/Automattic/newspack-plugin/assets/820752/3e5d3ab2-e239-4d03-ba20-05e36c04da70">

### How to test the changes in this Pull Request:

1. Check out this branch and https://github.com/Automattic/newspack-blocks/pull/1647
2. Confirm you have RAS enabled
3. On a fresh session, click on "Sign in" and confirm the modal renders as in the image above
4. Sign in with an existing reader account that has a configured password
5. Confirm you are prompted to enter the password and can successfully login
6. Click "Continue" and confirm you are redirected to the "My Account" page
7. On another fresh session click to Sign In and attempt to login with a new email
8. Confirm you get an error and clicking the "Create an account" link in the error message changes the form state
9. Register with the new email, confirm you're signed in and get the "Successful registration" as represented in the image above
10. Click to "Set a password" and confirm the email is sent and you are redirected to the My Account page with a notification
11. Before setting a password, start another fresh session and sign in with the newly registered email
12. Confirm you're prompted the enter the OTP 
13. Enter the OTP and confirm you sign in
14. Set a password for the account, start a fresh session and sign in
15. Confirm you're now prompted to set the password instead of an OTP

#### `openAuthModal( config )``

1. Make sure you have https://github.com/Automattic/newspack-blocks/pull/1647
2. On a fresh session, click to Donate using the donate block
3. Confirm you are prompted to authenticate
4. Authenticate and confirm the modal checkout appears
5. Close the modal and click to donate again
6. Confirm you're no longer prompted to authenticate and the modal checkout renders
7. Repeat the same steps using the "Checkout Button" block and confirm it behaves the same

### Other information:
